content/gophergala/data-mining.md

```diff
@@ -0,0 +1,154 @@
+A second worker aggregates the harvested data and stores it in the MongoDB database.
+The aggregated data is displayed through a simple Go http server.
+
+This post will focus on how data gets from chillingeffects.org into a MongoDB database via the harvester.
+
+## [chillingeffects.org](https://chillingeffects.org/)
+
+> The Chilling Effects database collects and analyzes legal complaints and requests for removal of online materials, helping Internet users to know their rights and understand the law. 
+> These data enable us to study the prevalence of legal threats and let Internet users see the source of content removals.
+
+# The chillingeffects.org API
+
+The [chillingeffects.org API](https://github.com/berkmancenter/chillingeffects/blob/master/doc/api_documentation.mkd) is a simple http JSON API.
+To harvest the Notices the harvester only requires one function of the API which is [request a notice](https://github.com/berkmancenter/chillingeffects/blob/master/doc/api_documentation.mkd#request-a-notice).
+The request is a GET call to the endpoint: ``https://chillingeffects.org/notices/<notice id>.json``
+On success the response body contains a JSON object.
+The package chillingeffects Go package has one function RequestNotice, which returns a Notice struct.
+
+# Simple sequential harvester
+
+Since bulk requests for notices are not possible with the API, each notice needs to be fetched on it's own.
+The simplest solution is to take an ID range and call the function RequestNotice for each ID.
+
+``` Go
+func harvestNotices(low, high int, session *mgo.Session) {
+	for id := low; id <= high; id++ {
+		notice, _ := chillingeffects.RequestNotice(id)
+		session.DB("").C("notices").Insert(notice)
+	}
+}
+```
+
+The problem is, it simply takes too long to fetch thousands of notices.
+Most of the time is spend waiting between a request and a response from chillingeffects.org and the database.
+If you use a worker service like iron.io and you are metered by the second your quota is exceeded very fast.
+
+
+# Infusing goroutines
+
+To reduce the time spend on each task, the harvester has to be optimized.
+One of Go's advertised features are coroutines called goroutines.
+A goroutine runs code concurrently to other goroutines.
+Since they have little overhead, a simple solution to the time problem would be to start each request in it's own goroutine.
+Which would look like this:
+
+``` Go
+func harvestNotices(low, high int, session *mgo.Session) {
+	for id := low; id <= high; id++ {
+		go function(id int, session *mgo.Session) {
+			notice, _ := chillingeffects.RequestNotice(id)
+			session.DB("").C("notices").Insert(notice)
+		}(id, session)
+	}
+}
+```
+
+The above code runs but most likely ends in an undesired result.
+The main goroutine which starts all the request goroutines finishes and ends the program before the other goroutines.
+The result is probably nothing but spend processing time.
```

Comment:
The program ends before the goroutines have a chance to requests and store the notices.